File: mobile/lib/models/plant.dart
class PlantCatalog {
  final String id;
  final String name;
  final String? nameLatin;
  final String plantType;
  final String? description;
  final bool isSystem;
  final List<PlantVariety> varieties;

  const PlantCatalog({
    required this.id,
    required this.name,
    this.nameLatin,
    required this.plantType,
    this.description,
    required this.isSystem,
    this.varieties = const [],
  });

  factory PlantCatalog.fromJson(Map<String, dynamic> json) => PlantCatalog(
        id: json['id'] as String,
        name: json['name'] as String,
        nameLatin: json['name_latin'] as String?,
        plantType: json['plant_type'] as String? ?? 'other',
        description: json['description'] as String?,
        isSystem: json['is_system'] as bool? ?? true,
        varieties: (json['varieties'] as List<dynamic>? ?? [])
            .map((e) => PlantVariety.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class PlantVariety {
  final String id;
  final String plantId;
  final String name;
  final String? description;

  const PlantVariety({
    required this.id,
    required this.plantId,
    required this.name,
    this.description,
  });

  factory PlantVariety.fromJson(Map<String, dynamic> json) => PlantVariety(
        id: json['id'] as String,
        plantId: json['plant_id'] as String,
        name: json['name'] as String,
        description: json['description'] as String?,
      );
}

class PlantInstance {
  final String id;
  final String zoneId;
  final String plantId;
  final String plantName;
  final String plantType;
  final String? varietyId;
  final String? varietyName;
  final String? customName;
  final DateTime? plantedDate;
  final int? quantity;
  final double? areaSqm;
  final String? notes;
  final String status;
  final DateTime createdAt;

  const PlantInstance({
    required this.id,
    required this.zoneId,
    required this.plantId,
    required this.plantName,
    required this.plantType,
    this.varietyId,
    this.varietyName,
    this.customName,
    this.plantedDate,
    this.quantity,
    this.areaSqm,
    this.notes,
    required this.status,
    required this.createdAt,
  });

  String get displayName =>
      customName ?? varietyName ?? plantName;

  String get statusLabel {
    const labels = {
      'planned':   'Запланировано',
      'active':    'Растёт',
      'harvested': 'Собран урожай',
      'removed':   'Удалено',
    };
    return labels[status] ?? status;
  }

  factory PlantInstance.fromJson(Map<String, dynamic> json) => PlantInstance(
        id:          json['id'] as String,
        zoneId:      json['zone_id'] as String,
        plantId:     json['plant_id'] as String,
        plantName:   json['plant_name'] as String,
        plantType:   json['plant_type'] as String? ?? 'other',
        varietyId:   json['variety_id'] as String?,
        varietyName: json['variety_name'] as String?,
        customName:  json['custom_name'] as String?,
        plantedDate: json['planted_date'] != null
            ? DateTime.parse(json['planted_date'] as String)
            : null,
        quantity:    json['quantity'] as int?,
        areaSqm:     (json['area_sqm'] as num?)?.toDouble(),
        notes:       json['notes'] as String?,
        status:      json['status'] as String? ?? 'active',
        createdAt:   DateTime.parse(json['created_at'] as String),
      );
}

File: backend/app/models/plant.py
import uuid
from datetime import datetime, timezone, date
from sqlalchemy import String, Boolean, DateTime, ForeignKey, Text, Integer, Date, Numeric
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID

from app.core.database import Base


class PlantCatalog(Base):
    """Справочник культур — Яблоня, Томат, Роза и т.д."""
    __tablename__ = "plant_catalog"

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)           # Яблоня
    name_latin: Mapped[str | None] = mapped_column(String(200))              # Malus domestica
    plant_type: Mapped[str] = mapped_column(String(50), nullable=False)      # tree/vegetable/...
    description: Mapped[str | None] = mapped_column(Text)
    days_to_harvest: Mapped[int | None] = mapped_column(Integer)
    sowing_months: Mapped[str | None] = mapped_column(String(50))           # "3,4,5"
    harvest_months: Mapped[str | None] = mapped_column(String(50))          # "8,9,10"
    is_custom: Mapped[bool] = mapped_column(Boolean, default=False)
    created_by: Mapped[str | None] = mapped_column(ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    varieties: Mapped[list["PlantVariety"]] = relationship(
        back_populates="plant", cascade="all, delete-orphan"
    )
    instances: Mapped[list["PlantInstance"]] = relationship(back_populates="plant")


class PlantVariety(Base):
    """Сорта культуры — Антоновка, Черри, Флорибунда и т.д."""
    __tablename__ = "plant_varieties"

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    plant_id: Mapped[str] = mapped_column(
        ForeignKey("plant_catalog.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)          # Антоновка
    description: Mapped[str | None] = mapped_column(Text)
    is_custom: Mapped[bool] = mapped_column(Boolean, default=False)
    created_by: Mapped[str | None] = mapped_column(ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    plant: Mapped["PlantCatalog"] = relationship(back_populates="varieties")
    instances: Mapped[list["PlantInstance"]] = relationship(back_populates="variety")


class PlantInstance(Base):
    """Экземпляр растения пользователя в конкретной зоне."""
    __tablename__ = "plant_instances"

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    zone_id: Mapped[str] = mapped_column(
        ForeignKey("zones.id", ondelete="CASCADE"), nullable=False
    )
    plant_id: Mapped[str] = mapped_column(
        ForeignKey("plant_catalog.id"), nullable=False
    )
    variety_id: Mapped[str | None] = mapped_column(
        ForeignKey("plant_varieties.id"), nullable=True
    )
    custom_name: Mapped[str | None] = mapped_column(String(200))            # «Томат у теплицы»
    planted_date: Mapped[date | None] = mapped_column(Date)
    quantity: Mapped[int | None] = mapped_column(Integer)
    area_sqm: Mapped[float | None] = mapped_column(Numeric(8, 2))
    notes: Mapped[str | None] = mapped_column(Text)
    status: Mapped[str] = mapped_column(String(20), default="active")       # planned/active/harvested/removed
    created_by: Mapped[str] = mapped_column(ForeignKey("users.id"), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    plant: Mapped["PlantCatalog"] = relationship(back_populates="instances")
    variety: Mapped["PlantVariety | None"] = relationship(back_populates="instances")
    zone: Mapped["Zone"] = relationship()  # noqa